lib/libgcov/gcov_gcc.c

```diff
@@ -0,0 +1,411 @@
+	/* If a dry run is required */
+	bool dry_run_done;
+} s_int_data;
+
+/*
+ * See gcov_impl_t->get_version
+ */
+static uint32_t gcov_gcc_get_version(void)
+{
+	return s_int_data.gcc_version;
+}
+
+/*
+ * See gcov_impl_t->init
+ */
+static void gcov_gcc_init(void *cov_object_info)
+{
+	struct object_data *od = cov_object_info;
+
+	/* The objects are inserted in LIFO */
+	if (!s_int_data.linked_list_objects) {
+		s_int_data.linked_list_objects = od;
+	} else {
+		od->next = s_int_data.linked_list_objects;
+		s_int_data.linked_list_objects = od;
+	}
+
+	if (!s_int_data.gcc_version)
+		s_int_data.gcc_version = od->version;
+}
+
+/*
+ * See gcov_impl_t->merge_add
+ */
+static void gcov_gcc_merge_add(void *ctrs, uint32_t nb_ctrs)
+{
+	(void)ctrs;
+	(void)nb_ctrs;
+}
+
+static void gcov_gcc_exit(void)
+{
+	/* empty */
+}
+
+/*
+ * See gcov_impl_t->reset_all_coverage_data
+ */
+static TEE_Result gcov_gcc_reset_all_coverage_data(void)
+{
+	struct object_data *od = s_int_data.linked_list_objects;
+	const struct func_data *fd = NULL;
+	const struct ctr_data *cd = NULL;
+	uint32_t cur_fn = 0;
+	uint32_t cur_ctr = 0;
+
+	/* No object registered */
+	if (!od)
+		return TEE_SUCCESS;
+
+	/* We loop over the object data */
+	do {
+		/* We loop over the function data */
+		for (cur_fn = 0; cur_fn < od->nb_functions; cur_fn++) {
+			fd = od->arr_func_data[cur_fn];
+
+			/* We loop over the type of counters */
+			for (cur_ctr = START_CTR_SUPPORTED;
+			     cur_ctr <= LAST_CTR_SUPPORTED; cur_ctr++) {
+				cd = &fd->ctrs_data[cur_ctr];
+
+				/* We reset the enabled counters */
+				if (od->merge_fn[cur_ctr])
+					memset(cd->ctrs, 0,
+					       sizeof(*cd->ctrs) * cd->nb_ctrs);
+			}
+		}
+	} while ((od = od->next));
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * See gcov_impl_t->start_dump_all_coverage_data
+ */
+static TEE_Result gcov_gcc_start_dump_all_coverage_data(void)
+{
+	/* Reuse the biggest buffer size if any */
+	uint32_t buf_size = s_int_data.dump_buf_size;
+
+	s_int_data.dump_buf = malloc(buf_size);
+	if (!s_int_data.dump_buf)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	s_int_data.dump_buf_size = buf_size;
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * We will always be aligned on a word
+ */
+#define write_u32(buf, val, dry_run) \
+	({ \
+		if (!dry_run) { \
+			*((uint32_t *)buf) = val; \
+			buf = ((uint8_t *)buf) + sizeof(uint32_t); \
+		} \
+		(sizeof(uint32_t)); \
+	})
+
+/*
+ * We are not always aligned on a word boundary else the use of memmove
+ */
+#define write_u64(buf, val, dry_run) \
+	({ \
+		if (!dry_run) { \
+			uint64_t v = val; \
+			memmove(buf, &v, sizeof(uint64_t)); \
+			buf = ((uint8_t *)buf) + sizeof(uint64_t); \
+		} \
+		(sizeof(uint64_t)); \
+	})
+
+/*
+ * Create a GCDA file in the buffer provided
+ *
+ * If the size of the buffer is zero, it returns the size required
+ *
+ * From gcc/gcc/gcov-dump.c
+ *
+ * A FUNCTION block is generally foloowed by blocks for counters, 1 bloc per
+ * counter
+ *
+ * struct gcda_file {
+ *     uint32_t magic;
```

Comment:
gcov-dump.c seems to generally not use fixed width integers. `unsigned int` is used instead of `uint32_t`

Reply:
The use of these values is to map as close to possible to the binary representation of the file.